src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -156,21 +158,44 @@ private void WatchResourceChanges(CancellationToken cancellationToken)
 
         async Task WatchKubernetesResource<T>(Func<WatchEventType, T, Task> handler) where T : CustomResource
         {
-            try
+            var retryUntilCancelled = new RetryStrategyOptions()
             {
-                await foreach (var (eventType, resource) in kubernetesService.WatchAsync<T>(cancellationToken: cancellationToken))
+                ShouldHandle = new PredicateBuilder().HandleInner<EndOfStreamException>(),
+                BackoffType = DelayBackoffType.Constant,
+                MaxRetryAttempts = int.MaxValue,
+                UseJitter = true,
+                MaxDelay = TimeSpan.FromSeconds(30),
+                OnRetry = (retry) =>
                 {
-                    await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    _logger.LogDebug(
+                        retry.Outcome.Exception,
+                        "Long poll watch operation was ended by server after {LongPollDurationInMs} milliseconds.",
+                        retry.Duration.TotalMilliseconds
+                        );
+                    return ValueTask.CompletedTask;
+                }
+            };
 
-                    try
-                    {
-                        await handler(eventType, resource).ConfigureAwait(false);
-                    }
-                    finally
+            var pipeline = new ResiliencePipelineBuilder().AddRetry(retryUntilCancelled).Build();
+
+            try
+            {
+                await pipeline.ExecuteAsync(async (pipelineCancellationToken) =>
```

Comment:
Is it much refactoring/cognitive overload to refactor this to pass the state through into Polly so you can avoid the need to allocation the closure here by making the lambda static, or is the context where this happens not perf-sensitive enough for that to be a concern?

Reply:
Its not that perf sensitive. We execute this once per resource type of project, executable, and container. And if we fail with an end of stream (after ~30 mins) its not a huge cost.